### PR TITLE
Allow private symbols to be control flow narrowed

### LINKED ISSUE #39960: Class property inference from constructor does not work with private class fields
**TypeScript Version:**  4.0.0-dev.20200803

**Search Terms:** class property inference, private class fields

**Expected behavior:** private class field's type is inferred from its assignment in the constructor

**Actual behavior:** typechecker reports implicit `any` for private class field 

**Related Issues:**

**Code**
```ts
class Example {
  #test;

  constructor(test: number) {
    this.#test = test;
  }

  get test() {
    return this.#test
  }
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, privateMap, value) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to set private field on non-instance");
    }
    privateMap.set(receiver, value);
    return value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, privateMap) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to get private field on non-instance");
    }
    return privateMap.get(receiver);
};
var _test;
class Example {
    constructor(test) {
        _test.set(this, void 0);
        __classPrivateFieldSet(this, _test, test);
    }
    get test() {
        return __classPrivateFieldGet(this, _test);
    }
}
_test = new WeakMap();

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ts=4.0.0-dev.20200803#code/MYGwhgzhAECiAeYC2AHEBTaBvAUNaAxAC7oREDcOe0wA9gHZkBOArsEbUwBQlkBc0eiyQAjdEwCU2aviIALAJYQAdMVJFoAXmi8K1AL5V8Ac3QbdXKbnz4mZlk3o7FKtWQM59QA)

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #39960

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
While I trust the change, I'm going to say that without more than a week of testing, we should probably just err on the safe side and make this 4.1-bound at best.

**Comment by weswigham** [comment]
Ping for reviewers - @DanielRosenwasser  and @ahejlsberg  - we're in the `4.1` era now~

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7925,7 +7925,7 @@ namespace ts {
         }
 
         function getFlowTypeInConstructor(symbol: Symbol, constructor: ConstructorDeclaration) {
-            const reference = factory.createPropertyAccessExpression(factory.createThis(), unescapeLeadingUnderscores(symbol.escapedName));
+            const reference = factory.createPropertyAccessExpression(factory.createThis(), startsWith(symbol.escapedName as string, "__#") ? factory.createPrivateIdentifier((symbol.escapedName as string).split("@")[1]) : unescapeLeadingUnderscores(symbol.escapedName));
```
Comment (sandersn) [inline src/compiler/checker.ts:7928]: please reformat this line to make it shorter

**Review by sandersn — APPROVED** [review]
LGTM, just one formatting request.

**Pull request merged**